File: mobile/android/app/src/main/kotlin/farm/ballydidean/weather/widget/WidgetForecast.kt
package farm.ballydidean.weather.widget

import java.time.Duration
import java.time.Instant
import java.time.LocalDate
import java.time.ZoneId
import java.time.ZoneOffset

enum class ForecastMode {
    ADJUSTED,
    RAW,
    UNAVAILABLE
}

enum class ForecastStatus {
    ADJUSTED,
    MIXED,
    RAW,
    UNAVAILABLE
}

data class ForecastSource(
    val runAt: Instant?,
    val receivedAt: Instant,
)

data class ForecastField(
    val mode: ForecastMode,
    val raw: Double?,
    val rawSource: ForecastSource?,
    val reason: String,
    val selected: Double?,
    val selectedSource: ForecastSource?,
    val selectedUntil: Instant?,
)

data class ForecastHour(
    val start: Instant,
    val end: Instant,
    val temperatureC: ForecastField,
    val rainMmPerHour: ForecastField,
    val cloudCoverPercent: ForecastField? = null,
    val windSpeedMps: ForecastField? = null,
)

data class ForecastCalendar(
    val date: LocalDate,
    val dayStart: Instant,
    val dayEnd: Instant,
    val cutoff: Instant,
    val sunset: Instant?,
    val overnightEnd: Instant? = null,
)

data class WidgetForecastSnapshot(
    val generatedAt: Instant,
    val receivedAt: Instant,
    val calendar: ForecastCalendar,
    val hours: List<ForecastHour>,
    val status: ForecastStatus,
)

object WidgetForecastDecoder {
    const val MAX_BYTES = 128 * 1024
    private val instantPattern = Regex("^[0-9]{4}-[0-9]{2}-[0-9]{2}T[0-9]{2}:[0-9]{2}:[0-9]{2}\\.[0-9]{3}Z$")
    private val datePattern = Regex("^[0-9]{4}-[0-9]{2}-[0-9]{2}$")
    private val siteZone = ZoneId.of("America/Los_Angeles")
    private val rootKeys = setOf("attribution", "calendar", "generatedAt", "hours", "receivedAt", "schemaVersion", "site", "status")
    private val fieldKeys = setOf("mode", "raw", "rawSource", "reason", "selected", "selectedSource", "selectedUntil")
    private val adjustedReasons = setOf("generic_adjustment", "independent_adjustment", "rain_adjustment")
    private val rawReasons = setOf("deadline_expired", "deadline_unavailable", "raw_forecast")

    // validate one complete public snapshot
    fun decode(bytes: ByteArray): WidgetForecastSnapshot {
        require(bytes.size <= MAX_BYTES) { "widget snapshot exceeds 128 KiB" }
        val root = StrictJson.parse(bytes).closedObject(rootKeys, "root")
        val version = root.string("schemaVersion")
        require(version in setOf("weather-widget/v1", "weather-widget/v2", "weather-widget/v3")) { "unsupported schema version" }
        val hasConditions = version != "weather-widget/v1"
        val hasOvernight = version == "weather-widget/v3"
        validateAttribution(root.objectValue("attribution"))
        validateSite(root.objectValue("site"))
        val generatedAt = instant(root.string("generatedAt"), "generatedAt")
        val receivedAt = instant(root.string("receivedAt"), "receivedAt")
        require(generatedAt <= receivedAt) { "snapshot clocks are reversed" }
        val calendar = decodeCalendar(root.objectValue("calendar"), generatedAt, hasOvernight)
        val hours = root.array("hours").mapIndexed { index, value ->
            decodeHour(value, generatedAt, calendar, index, hasConditions)
        }
        val horizon = calendar.overnightEnd ?: calendar.dayEnd
        require(hours.size.toLong() == Duration.between(calendar.dayStart, horizon).toHours()) { "hour count is outside the site calendar" }
        require(hours.lastOrNull()?.end == horizon) { "hour grid does not reach forecast end" }
        val status = status(root.string("status"))
        val derivedStatus = summarize(hours.flatMap { listOfNotNull(it.temperatureC.mode, it.rainMmPerHour.mode, it.cloudCoverPercent?.mode, it.windSpeedMps?.mode) })
        require(status == derivedStatus) { "snapshot status does not match fields" }
        return WidgetForecastSnapshot(generatedAt, receivedAt, calendar, hours, status)
    }

    // validate fixed public attribution
    private fun validateAttribution(value: JsonObject) {
        value.closedObject(setOf("label", "licenseUrl", "providerUrl"), "attribution")
        require(value.string("label") == "Open-Meteo · CC BY 4.0") { "unexpected attribution" }
        require(value.string("licenseUrl") == "https://creativecommons.org/licenses/by/4.0/") { "unexpected license" }
        require(value.string("providerUrl") == "https://open-meteo.com/") { "unexpected provider" }
    }

    // validate the one supported public site
    private fun validateSite(value: JsonObject) {
        value.closedObject(setOf("latitude", "longitude", "name", "slug", "timezone"), "site")
        require(value.string("name") == "Ballydidean") { "unexpected site name" }
        require(value.string("slug") == "ballydidean") { "unexpected site slug" }
        require(value.string("timezone") == siteZone.id) { "unexpected site timezone" }
        require(value.number("latitude") == 47.950429954185445) { "unexpected site latitude" }
        require(value.number("longitude") == -122.42797012608193) { "unexpected site longitude" }
    }

    // validate generated-at anchored calendar fields
    private fun decodeCalendar(value: JsonObject, generatedAt: Instant, hasOvernight: Boolean): ForecastCalendar {
        // extend only the new closed contract
        val keys = setOf("cutoff", "date", "dayEnd", "dayStart", "sunset") + if (hasOvernight) setOf("overnightEnd") else emptySet()
        value.closedObject(keys, "calendar")
        val dateText = value.string("date")
        require(datePattern.matches(dateText)) { "calendar date is malformed" }
        val date = LocalDate.parse(dateText)
        val dayStart = instant(value.string("dayStart"), "dayStart")
        val dayEnd = instant(value.string("dayEnd"), "dayEnd")
        val cutoff = instant(value.string("cutoff"), "cutoff")
        val expectedDayStart = date.atStartOfDay(siteZone).toInstant()
        val expectedDayEnd = date.plusDays(1).atStartOfDay(siteZone).toInstant()
        val expectedCutoff = date.atTime(20, 0).atZone(siteZone).toInstant()
        require(dayStart == expectedDayStart && dayEnd == expectedDayEnd && cutoff == expectedCutoff) {
            "calendar bounds do not match the site date"
        }
        // retain the preceding evening until seven in the site timezone
        val overnightEnd = if (hasOvernight) instant(value.string("overnightEnd"), "overnightEnd") else null
        if (hasOvernight) {
            require(overnightEnd == date.plusDays(1).atTime(7, 0).atZone(siteZone).toInstant()) { "overnight end is not seven am" }
            require(generatedAt >= date.atTime(7, 0).atZone(siteZone).toInstant() && generatedAt < overnightEnd) {
                "generatedAt is outside the forecast day"
            }
        } else {
            require(generatedAt >= dayStart && generatedAt < dayEnd) { "generatedAt is outside the anchored day" }
        }
        val sunset = nullableInstant(value.values.getValue("sunset"), "sunset")
        return ForecastCalendar(date, dayStart, dayEnd, cutoff, sunset, overnightEnd)
    }

    // validate one exact hourly interval
    private fun decodeHour(
        value: JsonValue,
        generatedAt: Instant,
        calendar: ForecastCalendar,
        index: Int,
        hasConditions: Boolean,
    ): ForecastHour {
        val baseKeys = setOf("end", "rainMmPerHour", "start", "temperatureC")
        // keep both versioned hourly contracts closed
        val keys = if (hasConditions) baseKeys + setOf("cloudCoverPercent", "windSpeedMps") else baseKeys
        val objectValue = value.closedObject(keys, "hour")
        val start = instant(objectValue.string("start"), "hour start")
        val end = instant(objectValue.string("end"), "hour end")
        val expectedStart = calendar.dayStart.plus(Duration.ofHours(index.toLong()))
        require(start == expectedStart && end == start.plus(Duration.ofHours(1))) { "hour grid is not contiguous" }
        require(start >= calendar.dayStart && end <= (calendar.overnightEnd ?: calendar.dayEnd)) { "hour is outside the forecast horizon" }
        // retain absent v1 conditions without inventing sunny or calm weather
        val cloud = if (hasConditions) decodeField(objectValue.values.getValue("cloudCoverPercent"), generatedAt, 0.0, 100.0, "cloud") else null
        val wind = if (hasConditions) decodeField(objectValue.values.getValue("windSpeedMps"), generatedAt, 0.0, 150.0, "wind") else null
        require(cloud?.mode != ForecastMode.ADJUSTED) { "cloud corrections are unsupported" }
        require(wind?.mode != ForecastMode.ADJUSTED || wind.reason == "generic_adjustment") { "wind correction reason is invalid" }
        return ForecastHour(
            start,
            end,
            decodeField(objectValue.values.getValue("temperatureC"), generatedAt, -100.0, 70.0, "temperature"),
            decodeField(objectValue.values.getValue("rainMmPerHour"), generatedAt, 0.0, 2000.0, "rain"),
            cloud,
            wind,
        )
    }

    // validate one bounded value and its paired provenance
    private fun decodeField(
        value: JsonValue,
        generatedAt: Instant,
        minimum: Double,
        maximum: Double,
        label: String,
    ): ForecastField {
        val objectValue = value.closedObject(fieldKeys, label)
        val mode = mode(objectValue.string("mode"))
        val raw = nullableNumber(objectValue.values.getValue("raw"), minimum, maximum, "$label raw")
        val selected = nullableNumber(objectValue.values.getValue("selected"), minimum, maximum, "$label selected")
        val rawSource = nullableSource(objectValue.values.getValue("rawSource"), generatedAt, "$label rawSource")
        val selectedSource = nullableSource(objectValue.values.getValue("selectedSource"), generatedAt, "$label selectedSource")
        val selectedUntil = nullableInstant(objectValue.values.getValue("selectedUntil"), "$label selectedUntil")
        val reason = objectValue.string("reason")
        // enforce mode-specific pairings
        when (mode) {
            ForecastMode.ADJUSTED -> {
                require(selected != null && selectedSource != null && selectedUntil != null) { "adjusted $label is incomplete" }
                require((raw == null) == (rawSource == null)) { "adjusted raw $label provenance is unpaired" }
                require(reason in adjustedReasons) { "adjusted $label reason is invalid" }
                require(selectedUntil > generatedAt) { "adjusted $label deadline is not future" }
            }
            ForecastMode.RAW -> {
                require(raw != null && rawSource != null && selected == raw) { "raw $label is incomplete" }
                require(selectedSource == null && selectedUntil == null) { "raw $label has adjusted metadata" }
                require(reason in rawReasons) { "raw $label reason is invalid" }
            }
            ForecastMode.UNAVAILABLE -> {
                require(raw == null && rawSource == null && selected == null) { "unavailable $label has a value" }
                require(selectedSource == null && selectedUntil == null && reason == "missing") { "unavailable $label metadata is invalid" }
            }
        }
        return ForecastField(mode, raw, rawSource, reason, selected, selectedSource, selectedUntil)
    }

    // validate one source's causal clocks
    private fun nullableSource(value: JsonValue, generatedAt: Instant, label: String): ForecastSource? {
        // retain explicit null
        if (value === JsonNull) {
            return null
        }
        val objectValue = value.closedObject(setOf("receivedAt", "runAt"), label)
        val receivedAt = instant(objectValue.string("receivedAt"), "$label receivedAt")
        val runAt = nullableInstant(objectValue.values.getValue("runAt"), "$label runAt")
        require(receivedAt <= generatedAt) { "$label receipt is after generatedAt" }
        require(runAt == null || runAt <= receivedAt) { "$label clocks are reversed" }
        return ForecastSource(runAt, receivedAt)
    }

    // parse a bounded nullable number
    private fun nullableNumber(value: JsonValue, minimum: Double, maximum: Double, label: String): Double? {
        // retain explicit null
        if (value === JsonNull) {
            return null
        }
        val number = (value as? JsonNumber)?.value ?: throw IllegalArgumentException("$label must be numeric or null")
        require(number in minimum..maximum) { "$label is outside bounds" }
        return number
    }

    // parse a strict millisecond utc instant
    private fun instant(value: String, label: String): Instant {
        require(instantPattern.matches(value)) { "$label is malformed" }
        return try {
            Instant.parse(value)
        } catch (error: RuntimeException) {
            throw IllegalArgumentException("$label is malformed", error)
        }
    }

    // parse an instant or explicit null
    private fun nullableInstant(value: JsonValue, label: String): Instant? {
        // retain explicit null
        if (value === JsonNull) {
            return null
        }
        val text = (value as? JsonString)?.value ?: throw IllegalArgumentException("$label must be a string or null")
        return instant(text, label)
    }

    // map one public mode
    private fun mode(value: String): ForecastMode = when (value) {
        "adjusted" -> ForecastMode.ADJUSTED
        "raw" -> ForecastMode.RAW
        "unavailable" -> ForecastMode.UNAVAILABLE
        else -> throw IllegalArgumentException("unsupported forecast mode")
    }

    // map one public status
    private fun status(value: String): ForecastStatus = when (value) {
        "adjusted" -> ForecastStatus.ADJUSTED
        "mixed" -> ForecastStatus.MIXED
        "raw" -> ForecastStatus.RAW
        "unavailable" -> ForecastStatus.UNAVAILABLE
        else -> throw IllegalArgumentException("unsupported forecast status")
    }

    // derive the public aggregate mode
    internal fun summarize(modes: List<ForecastMode>): ForecastStatus {
        val unique = modes.toSet()
        // preserve one homogeneous mode
        if (unique.size == 1) {
            return when (unique.single()) {
                ForecastMode.ADJUSTED -> ForecastStatus.ADJUSTED
                ForecastMode.RAW -> ForecastStatus.RAW
                ForecastMode.UNAVAILABLE -> ForecastStatus.UNAVAILABLE
            }
        }
        return ForecastStatus.MIXED
    }
}

File: apps/web/src/widget-forecast.ts
import {
  eveningSunTimes,
  forecastMetricValue,
  fromSiteWallClock,
  parseForecastRecordsResponse,
  toSiteWallClock,
  type WeatherRecord,
} from "./index.js";

export const WIDGET_FORECAST_SCHEMA_VERSION = "weather-widget/v1" as const;
export const WIDGET_FORECAST_MAX_BYTES = 128 * 1_024;

export const WIDGET_FORECAST_SITE = {
  latitude: 47.950429954185445,
  longitude: -122.42797012608193,
  name: "Ballydidean",
  slug: "ballydidean",
  timezone: "America/Los_Angeles",
} as const;
const HOUR_MS = 3_600_000;
const ADJUSTMENT_MAX_AGE_MS = 90 * 60_000;
const RAIN_DECISION_MAX_AGE_MS = 12 * HOUR_MS;

export type WidgetForecastMode = "adjusted" | "raw" | "unavailable";
export type WidgetForecastReason =
  | "deadline_expired"
  | "deadline_unavailable"
  | "generic_adjustment"
  | "independent_adjustment"
  | "missing"
  | "rain_adjustment"
  | "raw_forecast";
export type WidgetForecastStatus = WidgetForecastMode | "mixed";

export interface WidgetForecastSource {
  readonly runAt: string | null;
  readonly receivedAt: string;
}

export interface WidgetForecastField {
  readonly mode: WidgetForecastMode;
  readonly raw: number | null;
  readonly rawSource: WidgetForecastSource | null;
  readonly reason: WidgetForecastReason;
  readonly selected: number | null;
  readonly selectedSource: WidgetForecastSource | null;
  readonly selectedUntil: string | null;
}

export interface WidgetForecastHour {
  readonly end: string;
  readonly rainMmPerHour: WidgetForecastField;
  readonly start: string;
  readonly temperatureC: WidgetForecastField;
}

export interface WidgetForecastSnapshot {
  readonly attribution: {
    readonly label: "Open-Meteo · CC BY 4.0";
    readonly licenseUrl: "https://creativecommons.org/licenses/by/4.0/";
    readonly providerUrl: "https://open-meteo.com/";
  };
  readonly calendar: {
    readonly cutoff: string;
    readonly date: string;
    readonly dayEnd: string;
    readonly dayStart: string;
    readonly sunset: string | null;
  };
  readonly generatedAt: string;
  readonly hours: readonly WidgetForecastHour[];
  readonly receivedAt: string;
  readonly schemaVersion: typeof WIDGET_FORECAST_SCHEMA_VERSION;
  readonly site: {
    readonly latitude: typeof WIDGET_FORECAST_SITE.latitude;
    readonly longitude: typeof WIDGET_FORECAST_SITE.longitude;
    readonly name: typeof WIDGET_FORECAST_SITE.name;
    readonly slug: typeof WIDGET_FORECAST_SITE.slug;
    readonly timezone: typeof WIDGET_FORECAST_SITE.timezone;
  };
  readonly status: WidgetForecastStatus;
}

interface ForecastEnvelope {
  readonly generatedAt: string;
  readonly site: {
    readonly latitude: number;
    readonly longitude: number;
    readonly name: string;
    readonly slug: string;
    readonly timezone: string;
  };
}

type FieldKind = "rain" | "temperature";

// retain one validated source envelope for versioned projections
export interface WidgetForecastProjectionContext {
  readonly generatedAt: string;
  readonly parsed: ReturnType<typeof parseForecastRecordsResponse>;
  readonly receivedAt: string;
}

// project one validated forecast response into the closed widget contract
export function projectWidgetForecast(
  value: unknown,
  receivedAtValue: string,
): WidgetForecastSnapshot {
  const context = parseWidgetForecastProjection(value, receivedAtValue);
  const date = toSiteWallClock(context.generatedAt, WIDGET_FORECAST_SITE.timezone).slice(0, 10);
  const nextDate = addWidgetForecastCalendarDays(date, 1);
  const dayStart = canonicalWidgetForecastInstant(
    fromSiteWallClock(`${date}T00:00`, WIDGET_FORECAST_SITE.timezone),
    "calendar.dayStart",
  );
  const dayEnd = canonicalWidgetForecastInstant(
    fromSiteWallClock(`${nextDate}T00:00`, WIDGET_FORECAST_SITE.timezone),
    "calendar.dayEnd",
  );
  const cutoff = canonicalWidgetForecastInstant(
    fromSiteWallClock(`${date}T20:00`, WIDGET_FORECAST_SITE.timezone),
    "calendar.cutoff",
  );
  const expectedStarts = widgetForecastHourlyInstants(dayStart, dayEnd, 23, 25);
  const records = indexWidgetForecastRecords(context.parsed.data, expectedStarts);
  const sunset = eveningSunTimes(WIDGET_FORECAST_SITE, new Date(context.generatedAt)).sunset;
  const hours = expectedStarts.map(
    // project each reviewed civil-day interval independently
    (start) => projectWidgetForecastHour(records.get(start) ?? null, start, context),
  );
  const snapshot: WidgetForecastSnapshot = {
    attribution: {
      label: "Open-Meteo · CC BY 4.0",
      licenseUrl: "https://creativecommons.org/licenses/by/4.0/",
      providerUrl: "https://open-meteo.com/",
    },
    calendar: {
      cutoff,
      date,
      dayEnd,
      dayStart,
      sunset: sunset?.toISOString() ?? null,
    },
    generatedAt: context.generatedAt,
    hours,
    receivedAt: context.receivedAt,
    schemaVersion: WIDGET_FORECAST_SCHEMA_VERSION,
    site: {
      latitude: WIDGET_FORECAST_SITE.latitude,
      longitude: WIDGET_FORECAST_SITE.longitude,
      name: WIDGET_FORECAST_SITE.name,
      slug: WIDGET_FORECAST_SITE.slug,
      timezone: WIDGET_FORECAST_SITE.timezone,
    },
    status: summarizeStatus(hours),
  };

  // enforce the public response ceiling before the edge adapter is added
  if (new TextEncoder().encode(JSON.stringify(snapshot)).byteLength > WIDGET_FORECAST_MAX_BYTES) {
    throw new RangeError("Widget forecast exceeds 128 KiB");
  }

  return snapshot;
}

// validate one upstream response without choosing a versioned calendar
export function parseWidgetForecastProjection(
  value: unknown,
  receivedAtValue: string,
): WidgetForecastProjectionContext {
  const envelope = parseEnvelope(value);
  const generatedAt = canonicalWidgetForecastInstant(envelope.generatedAt, "generatedAt");
  const receivedAt = canonicalWidgetForecastInstant(receivedAtValue, "receivedAt");

  // reject a forecast anchor the edge could not yet have received
  if (Date.parse(generatedAt) > Date.parse(receivedAt)) {
    throw new RangeError("Widget forecast generatedAt is after receivedAt");
  }

  return {
    generatedAt,
    parsed: parseForecastRecordsResponse(value),
    receivedAt,
  };
}

// project one genuine forecast row through the shared v1 field rules
export function projectWidgetForecastHour(
  record: WeatherRecord | null,
  start: string,
  context: WidgetForecastProjectionContext,
): WidgetForecastHour {
  return {
    end: new Date(Date.parse(start) + HOUR_MS).toISOString(),
    rainMmPerHour: projectField(
      record,
      "rain",
      context.generatedAt,
      context.parsed.adjustmentRuntime,
      context.parsed.temperatureAdjustmentRuntime,
    ),
    start,
    temperatureC: projectField(
      record,
      "temperature",
      context.generatedAt,
      context.parsed.adjustmentRuntime,
      context.parsed.temperatureAdjustmentRuntime,
    ),
  };
}

// validate the anchor and exact product site before invoking the shared parser
function parseEnvelope(value: unknown): ForecastEnvelope {
  const envelope = objectValue(value);
  const site = objectValue(envelope?.site);

  // require the sole calendar anchor and allowlisted site identity
  if (
    envelope === null ||
    typeof envelope.generatedAt !== "string" ||
    site === null ||
    site.slug !== WIDGET_FORECAST_SITE.slug ||
    site.timezone !== WIDGET_FORECAST_SITE.timezone ||
    site.latitude !== WIDGET_FORECAST_SITE.latitude ||
    site.longitude !== WIDGET_FORECAST_SITE.longitude ||
    site.name !== WIDGET_FORECAST_SITE.name
  ) {
    throw new RangeError("Widget forecast site or calendar anchor is invalid");
  }

  return envelope as unknown as ForecastEnvelope;
}

// narrow one untrusted json object
function objectValue(value: unknown): Record<string, unknown> | null {
  // reject arrays and primitives
  if (value === null || typeof value !== "object" || Array.isArray(value)) {
    return null;
  }

  return value as Record<string, unknown>;
}

// normalize one explicit-zone instant
export function canonicalWidgetForecastInstant(value: string, field: string): string {
  const match = /^(?<year>\d{4})-(?<month>\d{2})-(?<day>\d{2})T(?<hour>\d{2}):(?<minute>\d{2}):(?<second>\d{2})(?:\.\d{1,3})?(?<zone>Z|[+-]\d{2}:\d{2})$/u.exec(value);

  // reject local, normalized, and malformed timestamps
  if (match?.groups === undefined || !Number.isFinite(Date.parse(value))) {
    throw new RangeError(`Widget forecast ${field} is invalid`);
  }

  const year = Number(match.groups.year);
  const month = Number(match.groups.month);
  const day = Number(match.groups.day);
  const hour = Number(match.groups.hour);
  const minute = Number(match.groups.minute);
  const second = Number(match.groups.second);
  const zone = match.groups.zone!;
  const offsetHour = zone === "Z" ? 0 : Number(zone.slice(1, 3));
  const offsetMinute = zone === "Z" ? 0 : Number(zone.slice(4, 6));
  const monthLength = month >= 1 && month <= 12
    ? new Date(Date.UTC(year, month, 0)).getUTCDate()
    : 0;

  // reject every calendar and zone component overflow
  if (
    day < 1 ||
    day > monthLength ||
    hour > 23 ||
    minute > 59 ||
    second > 59 ||
    offsetHour > 23 ||
    offsetMinute > 59
  ) {
    throw new RangeError(`Widget forecast ${field} is invalid`);
  }

  return new Date(value).toISOString();
}

// advance one iso calendar date without applying a runtime timezone
export function addWidgetForecastCalendarDays(date: string, days: number): string {
  const [year, month, day] = date.split("-").map(Number);

  // reject an unexpected wall-clock representation
  if (year === undefined || month === undefined || day === undefined) {
    throw new RangeError("Widget forecast calendar date is invalid");
  }

  return new Date(Date.UTC(year, month - 1, day + days)).toISOString().slice(0, 10);
}

// enumerate every real utc hour in one site-local day
export function widgetForecastHourlyInstants(
  dayStart: string,
  dayEnd: string,
  minimumHours: number,
  maximumHours: number,
): readonly string[] {
  const start = Date.parse(dayStart);
  const end = Date.parse(dayEnd);
  const count = (end - start) / HOUR_MS;

  // accept only reviewed civil-day capacities
  if (!Number.isSafeInteger(count) || count < minimumHours || count > maximumHours) {
    throw new RangeError("Widget forecast day has an unsupported hour count");
  }

  return Array.from(
    { length: count },
    // retain explicit repeated-hour instants during fall-back
    (_, index) => new Date(start + index * HOUR_MS).toISOString(),
  );
}

// reject duplicate and foreign-day rows while allowing genuine missing hours
export function indexWidgetForecastRecords(
  records: readonly WeatherRecord[],
  expectedStarts: readonly string[],
): ReadonlyMap<string, WeatherRecord> {
  const allowed = new Set(expectedStarts);
  const indexed = new Map<string, WeatherRecord>();

  // validate each supplied forecast row once
  for (const record of records) {
    const start = canonicalWidgetForecastInstant(record.validAt, "hour.start");

    // reject nonforecast, off-day, and duplicate rows
    if (
      record.provenance.sourceKind !== "forecast" ||
      !allowed.has(start) ||
      indexed.has(start)
    ) {
      throw new RangeError("Widget forecast contains a duplicate or foreign hour");
    }

    indexed.set(start, record);
  }

  return indexed;
}

// project one field with its independent correction deadline and provenance
function projectField(
  record: WeatherRecord | null,
  kind: FieldKind,
  generatedAt: string,
  genericRuntime: ReturnType<typeof parseForecastRecordsResponse>["adjustmentRuntime"],
  temperatureRuntime: ReturnType<typeof parseForecastRecordsResponse>["temperatureAdjustmentRuntime"],
): WidgetForecastField {
  // preserve a missing grid row without inventing dry weather
  if (record === null) {
    return unavailableField("missing");
  }

  // reject either supplied raw rain representation when physically invalid
  if (
    kind === "rain" &&
    ((record.metrics.precipitationRateMmPerHour !== null &&
      !validFieldValue("rain", record.metrics.precipitationRateMmPerHour)) ||
      (record.metrics.precipitationMm !== null &&
        !validFieldValue("rain", record.metrics.precipitationMm)))
  ) {
    throw new RangeError("Widget forecast contains invalid rain");
  }

  const raw = fieldValue(record, kind, false);
  const validatedRawSource = recordSource(
    record.productRunAt,
    record.receivedAt,
    generatedAt,
  );

  // reject present field-specific physical violations for the whole snapshot
  if (raw !== null && !validFieldValue(kind, raw)) {
    throw new RangeError(`Widget forecast contains invalid ${kind}`);
  }

  const rawSource = raw === null ? null : validatedRawSource;
  const selected = fieldValue(record, kind, true);
  const adjustment = adjustmentSelection(
    record,
    kind,
    generatedAt,
    genericRuntime,
    temperatureRuntime,
  );

  // retain a truthful raw selection when no active correction survived parsing
  if (adjustment === null) {
    return raw === null
      ? unavailableField("missing")
      : rawField(raw, validatedRawSource, "raw_forecast");
  }

  // demote unbounded corrections to their captured raw counterpart
  if (adjustment.deadline === null) {
    return raw === null
      ? unavailableField("missing")
      : rawField(raw, validatedRawSource, "deadline_unavailable");
  }

  // demote corrections already expired at the snapshot anchor
  if (Date.parse(adjustment.deadline) <= Date.parse(generatedAt)) {
    return raw === null
      ? unavailableField("missing")
      : rawField(raw, validatedRawSource, "deadline_expired");
  }

  // reject an invalid correction at the public boundary
  if (!validFieldValue(kind, selected)) {
    throw new RangeError(`Widget forecast contains invalid selected ${kind}`);
  }

  return {
    mode: "adjusted",
    raw,
    rawSource,
    reason: adjustment.reason,
    selected,
    selectedSource: adjustment.source,
    selectedUntil: adjustment.deadline,
  };
}

// identify the active resolver branch without exposing private adjustment metadata
function adjustmentSelection(
  record: WeatherRecord,
  kind: FieldKind,
  generatedAt: string,
  genericRuntime: ReturnType<typeof parseForecastRecordsResponse>["adjustmentRuntime"],
  temperatureRuntime: ReturnType<typeof parseForecastRecordsResponse>["temperatureAdjustmentRuntime"],
): Readonly<{
  deadline: string | null;
  reason: Extract<WidgetForecastReason, "generic_adjustment" | "independent_adjustment" | "rain_adjustment">;
  source: WidgetForecastSource;
}> | null {
  const maximum = Date.parse(generatedAt) + ADJUSTMENT_MAX_AGE_MS;

  // prefer the independently selected temperature correction
  if (kind === "temperature" && record.temperatureAdjustment?.state === "active") {
    const source = record.temperatureAdjustment.sourceForecast;
    return {
      deadline: boundedDeadline(maximum, temperatureRuntime.expiresAt),
      reason: "independent_adjustment",
      source: recordSource(
        source?.runInitializedAt ?? null,
        source?.firstReceivedAt,
        generatedAt,
      ),
    };
  }

  // prefer the independently selected rain correction
  if (kind === "rain" && record.rainAdjustment?.state === "active") {
    const source = record.rainAdjustment.sourceForecast;
    return {
      deadline: rainDeadline(maximum, source?.decisionAt, generatedAt),
      reason: "rain_adjustment",
      source: recordSource(
        source?.runInitializedAt ?? null,
        source?.firstReceivedAt,
        generatedAt,
      ),
    };
  }

  // retain only the generic correction branch the shared resolver can select
  if (
    kind === "temperature" &&
    record.adjustment?.state === "active" &&
    record.adjustment.appliedMetrics.includes("temperatureC")
  ) {
    return {
      deadline: boundedDeadline(maximum, genericRuntime.expiresAt),
      reason: "generic_adjustment",
      source: recordSource(record.productRunAt, record.receivedAt, generatedAt),
    };
  }

  return null;
}

// resolve rain rate before its hourly-amount compatibility fallback
function fieldValue(
  record: WeatherRecord,
  kind: FieldKind,
  useAdjustments: boolean,
): number | null {
  // preserve temperature's single canonical metric
  if (kind === "temperature") {
    return forecastMetricValue(record, "temperatureC", useAdjustments);
  }

  const rate = forecastMetricValue(record, "precipitationRateMmPerHour", useAdjustments);

  // reject a present invalid rate rather than hiding it behind a valid amount
  if (rate !== null && !validFieldValue("rain", rate)) {
    return rate;
  }

  return rate ?? forecastMetricValue(record, "precipitationMm", useAdjustments);
}

// cap one known adjustment deadline at snapshot age
function boundedDeadline(maximum: number, value: string | null | undefined): string | null {
  // fail raw when the applicable expiry is absent
  if (value === null || value === undefined) {
    return null;
  }

  try {
    const canonical = canonicalWidgetForecastInstant(value, "adjustment deadline");
    return new Date(Math.min(maximum, Date.parse(canonical))).toISOString();
  } catch {
    // fail raw on a malformed required deadline
    return null;
  }
}

// derive rain expiry without exposing its private decision clock
function rainDeadline(
  maximum: number,
  value: string | undefined,
  generatedAt: string,
): string | null {
  // fail raw when the parsed decision unexpectedly lacks its clock
  if (value === undefined) {
    return null;
  }

  let decisionAt: string;

  try {
    decisionAt = canonicalWidgetForecastInstant(value, "rain decisionAt");
  } catch {
    // fail raw on a malformed required decision clock
    return null;
  }

  // reject decisions that had not happened at snapshot generation
  if (Date.parse(decisionAt) > Date.parse(generatedAt)) {
    throw new RangeError("Widget forecast rain decision is from the future");
  }

  return new Date(Math.min(
    maximum,
    Date.parse(decisionAt) + RAIN_DECISION_MAX_AGE_MS,
  )).toISOString();
}

// project one allowlisted source clock pair
function recordSource(
  runAt: string | null,
  receivedAt: string | undefined,
  generatedAt: string,
): WidgetForecastSource {
  // reject selected sources without a trustworthy receipt clock
  if (receivedAt === undefined) {
    throw new RangeError("Widget forecast selected source is incomplete");
  }

  const canonicalRunAt = runAt === null
      ? null
      : canonicalWidgetForecastInstant(runAt, "source.runAt");
  const canonicalReceivedAt = canonicalWidgetForecastInstant(receivedAt, "source.receivedAt");

  // reject impossible source ordering before it can produce negative ages
  if (
    Date.parse(canonicalReceivedAt) > Date.parse(generatedAt) ||
    (canonicalRunAt !== null && Date.parse(canonicalRunAt) > Date.parse(canonicalReceivedAt))
  ) {
    throw new RangeError("Widget forecast source timestamps are out of order");
  }

  return {
    runAt: canonicalRunAt,
    receivedAt: canonicalReceivedAt,
  };
}

// enforce canonical physical ranges at the public projection boundary
function validFieldValue(kind: FieldKind, value: number | null): value is number {
  // preserve explicit missing values
  if (value === null || !Number.isFinite(value)) {
    return false;
  }

  return kind === "temperature"
    ? value >= -100 && value <= 70
    : value >= 0 && value <= 2_000;
}

// create one raw field with matching raw and selected provenance
function rawField(
  raw: number,
  source: WidgetForecastSource,
  reason: Extract<WidgetForecastReason, "deadline_expired" | "deadline_unavailable" | "raw_forecast">,
): WidgetForecastField {
  return {
    mode: "raw",
    raw,
    rawSource: source,
    reason,
    selected: raw,
    selectedSource: null,
    selectedUntil: null,
  };
}

// create one unavailable field without counterfeit provenance
function unavailableField(
  reason: Extract<WidgetForecastReason, "missing">,
): WidgetForecastField {
  return {
    mode: "unavailable",
    raw: null,
    rawSource: null,
    reason,
    selected: null,
    selectedSource: null,
    selectedUntil: null,
  };
}

// summarize the complete anchored day without pretending gaps are dry
function summarizeStatus(hours: readonly WidgetForecastHour[]): WidgetForecastStatus {
  const modes = new Set<WidgetForecastMode>();

  // include both independently expiring fields
  for (const hour of hours) {
    modes.add(hour.temperatureC.mode);
    modes.add(hour.rainMmPerHour.mode);
  }

  // preserve a single honest mode when the whole day agrees
  if (modes.size === 1) {
    return [...modes][0]!;
  }

  return "mixed";
}
